### PR TITLE
Add transaction and ssl support for Artemis Connector

### PR DESCRIPTION
**Pull request by riyafa**
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/14776 https://github.com/ballerina-platform/ballerina-lang/issues/15159

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Required Balo version update
- [ ] Updated Change Log
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [x] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [x] Ballerina By Examples

### CODE REVIEW

File: examples/artemis-transaction-consumer/artmeis_transaction_consumer.description [hunk 1]
```diff
@@ -0,0 +1,2 @@
+// This example shows how transactions can be used during reception of a message in Artemis connector.
```
Comment (praneesha) [inline examples/artemis-transaction-consumer/artmeis_transaction_consumer.description]:
```suggestion
// This example shows how transactions can be used during the reception of a message in the Artemis connector.
```

File: examples/artemis-transaction-producer/artmeis_transaction_producer.description [hunk 2]
```diff
@@ -0,0 +1,2 @@
+// This example shows how transactions can be used during sending of a message in Artemis connector.
```
Comment (praneesha) [inline examples/artemis-transaction-producer/artmeis_transaction_producer.description]:
```suggestion
// This example shows how transactions can be used during the sending of a message in the Artemis connector.
```

File: examples/artemis-transaction-producer/artmeis_transaction_producer.description [hunk 3]
```diff
@@ -0,0 +1,2 @@
...
+// If the transaction fails due to an error or by the use of `abort` message sends will not be committed.
```
Comment (praneesha) [inline examples/artemis-transaction-producer/artmeis_transaction_producer.description]:
```suggestion
// If the transaction fails due to an error or due to the use of `abort`, the message will not be sent.
```

File: examples/artemis-transaction-consumer/artemis_transaction_consumer.bal [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import ballerina/artemis;
+import ballerina/io;
+
+public function main() returns error? {
+    // Create the Artemis listener with `autoCommitAcks` set to `false`.
```
Comment (praneesha) [inline examples/artemis-transaction-consumer/artemis_transaction_consumer.bal]:
```suggestion
    // Create the Artemis listener with the `autoCommitAcks` property set to `false`.
```

File: examples/artemis-transaction-consumer/artemis_transaction_consumer.bal [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    // In this case the message acknowledgements will be committed only
```
Comment (praneesha) [inline examples/artemis-transaction-consumer/artemis_transaction_consumer.bal]:
```suggestion
    // In this case, the message acknowledgments will be committed only
```

File: examples/artemis-transaction-consumer/artemis_transaction_consumer.bal [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    // if the transaction block completes successfully.
+    artemis:Listener lis = new artemis:Listener({host:"localhost", port:61616});
+    // Create a `Consumer` object to make blocking receive calls
```
Comment (praneesha) [inline examples/artemis-transaction-consumer/artemis_transaction_consumer.bal]:
```suggestion
    // Create a `Consumer` object to block receiving calls.
```
Is that what needs to be meant? The idea is not clear.

File: examples/artemis-transaction-consumer/artemis_transaction_consumer.bal [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    var consumer = lis.createAndGetConsumer({
+        queueName: "example"
+    });
+    if (consumer is artemis:Consumer) {
+        transaction {
+            // Makes a blocking receive call
```
Comment (praneesha) [inline examples/artemis-transaction-consumer/artemis_transaction_consumer.bal]:
```suggestion
            // Makes a blocking receive call
```

+1 if "blocking receive call" is a known technical term. If not, we will have to reword this.

File: examples/artemis-transaction-producer/artemis_transaction_producer.bal [hunk 8]
```diff
@@ -0,0 +1,23 @@
+import ballerina/artemis;
+import ballerina/io;
+
+public function main() {
+    // Create the Artemis `Producer` with `autoCommitSends` set to `false`.
```
Comment (praneesha) [inline examples/artemis-transaction-producer/artemis_transaction_producer.bal]:
```suggestion
    // Create the Artemis `Producer` with the `autoCommitSends` property set to `false`.
```

File: examples/artemis-transaction-producer/artemis_transaction_producer.bal [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    // In this case the message sends will be committed only
```
Comment (praneesha) [inline examples/artemis-transaction-producer/artemis_transaction_producer.bal]:
```suggestion
    // In this case, the message will be sent only
```

File: examples/artemis-transaction-producer/artemis_transaction_producer.bal [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+    // if the transaction block completes successfully.
+    artemis:Producer prod = new({host:"localhost", port:61616}, "example");
+    transaction {
+        // Sends a message to the queue inside the transaction block
```
Comment (praneesha) [inline examples/artemis-transaction-producer/artemis_transaction_producer.bal]:
```suggestion
        // Sends a message to the queue inside the transaction block.
```

File: examples/artemis-transaction-producer/artemis_transaction_producer.bal [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+        var err = prod->send("Hello from Ballerina!");
+        if (err is error) {
+            io:println("Error occurred sending message");
+        }
+    }
+    // Closes the producer
```
Comment (praneesha) [inline examples/artemis-transaction-producer/artemis_transaction_producer.bal]:
```suggestion
    // Close the producer.
```
For the verbs in the comments, be consistent in using either of the singular form or the plural form of the verb E.g., "Close" vs. "Closes".

File: stdlib/messaging/activemq-artemis/src/main/ballerina/artemis/connection.bal [hunk 12]
```diff
@@ -38,10 +38,91 @@ public type Connection client object {
...
+    }
+
+    private function getQueryParamPart(string key, string value) returns string {
+        return "&" + key + "=" + value;
+    }
 };
 
 # Configurations related to a Artemis `Connection`.
 #
+# + secureSocket - SSL/TLS related configs
```
Comment (praneesha) [inline stdlib/messaging/activemq-artemis/src/main/ballerina/artemis/connection.bal]:
```suggestion
# + secureSocket - Configurations related to SSL/TLS
```

File: stdlib/messaging/activemq-artemis/src/main/ballerina/artemis/listener.bal [hunk 13]
```diff
@@ -41,22 +47,33 @@ public type Listener object {
         return self.createConsumer(serviceType);
     }
 
-    function start() returns error? = external;
-    function createConsumer(service serviceType) returns error? = external;
-    function stop() returns error? = external;
+    # The Artemis consumer is represented by the Service object.
```
Comment (praneesha) [inline stdlib/messaging/activemq-artemis/src/main/ballerina/artemis/listener.bal]:
```suggestion
    # The Artemis consumer is represented by the `Service` object. 
```

File: stdlib/messaging/activemq-artemis/src/main/ballerina/artemis/listener.bal [hunk 14]
```diff
@@ -41,22 +47,33 @@ public type Listener object {
         return self.createConsumer(serviceType);
     }
 
-    function start() returns error? = external;
-    function createConsumer(service serviceType) returns error? = external;
-    function stop() returns error? = external;
+    # The Artemis consumer is represented by the Service object. 
+    # But if there is a need to make blocking recieve calls, probably within a transaction block
```
Comment (praneesha) [inline stdlib/messaging/activemq-artemis/src/main/ballerina/artemis/listener.bal]:
```suggestion
    # However, if there is a need to make blocking receive calls,  within a transaction block,
```
+1 if "blocking receive calls" is a known technical term.

File: stdlib/messaging/activemq-artemis/src/main/ballerina/artemis/listener.bal [hunk 15]
```diff
@@ -41,22 +47,33 @@ public type Listener object {
...
     }
 
-    function start() returns error? = external;
-    function createConsumer(service serviceType) returns error? = external;
-    function stop() returns error? = external;
+    # The Artemis consumer is represented by the Service object. 
+    # But if there is a need to make blocking recieve calls, probably within a transaction block 
+    # this method can be used to create and get a `Consumer` object.
+    # 
+    # + config - the configuration for the queue to consume from
```
Comment (praneesha) [inline stdlib/messaging/activemq-artemis/src/main/ballerina/artemis/listener.bal]:
```suggestion
    # + config - the configuration consumed by the queue
```

File: stdlib/messaging/activemq-artemis/src/main/ballerina/artemis/listener.bal [hunk 16]
```diff
@@ -41,22 +47,33 @@ public type Listener object {
...
+    # + autoAck - whether to automatically acknowledge the received message
```
Comment (praneesha) [inline stdlib/messaging/activemq-artemis/src/main/ballerina/artemis/listener.bal]:
```suggestion
    # + autoAck - whether to acknowledge the received message automatically
```

File: stdlib/messaging/activemq-artemis/src/main/ballerina/artemis/listener.bal [hunk 17]
```diff
@@ -41,22 +47,33 @@ public type Listener object {
...
-    function start() returns error? = external;
-    function createConsumer(service serviceType) returns error? = external;
-    function stop() returns error? = external;
+    # The Artemis consumer is represented by the Service object. 
+    # But if there is a need to make blocking recieve calls, probably within a transaction block 
+    # this method can be used to create and get a `Consumer` object.
+    # 
+    # + config - the configuration for the queue to consume from
+    # + autoAck - whether to automatically acknowledge the received message    
+    # + filter - only messages which match this filter will be consumed
```
Comment (praneesha) [inline stdlib/messaging/activemq-artemis/src/main/ballerina/artemis/listener.bal]:
```suggestion
    # + filter - only messages, which match this filter will be consumed
```

File: stdlib/messaging/activemq-artemis/src/main/ballerina/artemis/listener.bal [hunk 18]
```diff
@@ -41,22 +47,33 @@ public type Listener object {
...
+
+    private function start() returns error? = external;
+    private function createConsumer(service serviceType) returns error? = external;
+    private function stop() returns error? = external;
 };
 
 # The configuration for an Artemis consumer service.
 #
-# + autoAck - whether to automatically acknowledge a service when a resource completes
+# + autoAck - whether to automatically acknowledge a message when a resource completes
```
Comment (praneesha) [inline stdlib/messaging/activemq-artemis/src/main/ballerina/artemis/listener.bal]:
```suggestion
# + autoAck - whether to acknowledge a message automatically when a resource completes
```

File: stdlib/messaging/activemq-artemis/src/main/ballerina/artemis/session.bal [hunk 19]
```diff
@@ -42,7 +43,11 @@ public type Session client object {
 #
 # + username - The username
 # + password - The password
+# + autoCommitSends - Whether the transaction send is automatically committed
```
Comment (praneesha) [inline stdlib/messaging/activemq-artemis/src/main/ballerina/artemis/session.bal]:
```suggestion
# + autoCommitSends - Whether the transaction is automatically sent
```

File: stdlib/messaging/activemq-artemis/src/main/ballerina/artemis/session.bal [hunk 20]
```diff
@@ -42,7 +43,11 @@ public type Session client object {
...
+# + autoCommitAcks - Whether the transaction reception ack is automatically committed
```
Comment (praneesha) [inline stdlib/messaging/activemq-artemis/src/main/ballerina/artemis/session.bal]:
```suggestion
# + autoCommitAcks - Whether the transaction reception is acknowledged automatically 
```

File: examples/artemis-transaction-consumer/artemis_transaction_consumer.bal (hunk 5 above)
Comment (riyafa) [reply examples/artemis-transaction-consumer/artemis_transaction_consumer.bal]: I guess the correct spelling is acknowledgements

File: examples/artemis-transaction-consumer/artemis_transaction_consumer.bal (hunk 6 above)
Comment (riyafa) [reply examples/artemis-transaction-consumer/artemis_transaction_consumer.bal]: Meaning of blocking receive calls is "synchronous receive calls". I have modified the comments

File: examples/artemis-transaction-consumer/artemis_transaction_consumer.bal (hunk 7 above)
Comment (riyafa) [reply examples/artemis-transaction-consumer/artemis_transaction_consumer.bal]: changed to "synchronous receive"

File: examples/artemis-transaction-producer/artemis_transaction_producer.bal (hunk 9 above)
Comment (riyafa) [reply examples/artemis-transaction-producer/artemis_transaction_producer.bal]: It is that we are committing sends. This is for transactions.

File: examples/artemis-transaction-producer/artemis_transaction_producer.bal (hunk 11 above)
Comment (riyafa) [reply examples/artemis-transaction-producer/artemis_transaction_producer.bal]: Using closes like the sends in other places

File: stdlib/messaging/activemq-artemis/src/main/ballerina/artemis/listener.bal (hunk 15 above)
Comment (riyafa) [reply stdlib/messaging/activemq-artemis/src/main/ballerina/artemis/listener.bal]: It is the config for the queue not config consumed by queue :-1:

File: stdlib/messaging/activemq-artemis/src/main/ballerina/artemis/listener.bal [hunk 21]
```diff
@@ -19,20 +19,26 @@
...
+            }
+            self.session = new(connection, config = { username: sessionOrEndpointConfig["username"],
+                    password: sessionOrEndpointConfig["password"],
+                    autoCommitSends: false, autoCommitAcks:  false });
+            self.session.anonymousSession = true;
         }
     }
     public function __start() returns error? {
-        return self.start();
+        return self. start();
```
Comment (daneshk) [inline stdlib/messaging/activemq-artemis/src/main/ballerina/artemis/listener.bal]: Shall we remove additional space here

Comment (riyafa) [reply stdlib/messaging/activemq-artemis/src/main/ballerina/artemis/listener.bal]: This is an issue with VSCode formatting. Created an [issue](https://github.com/ballerina-platform/ballerina-lang/issues/15264) and I have fixed it manually here.

File: stdlib/messaging/activemq-artemis/src/main/java/org/ballerinalang/messaging/artemis/ArtemisTransactionContext.java [hunk 22]
```diff
@@ -77,21 +77,21 @@ public XAResource getXAResource() {
...
                 }
             }
         } else {
             if (!context.isInTransaction()) {
-                throw new BallerinaException(
-                        "Artemis transacted session object methods should be used inside a transaction block ",
-                        context);
+                context.setError(ArtemisUtils.getError(context, "The Session used by the Artemis " + objectType +
+                        " object is transacted. Hence " + objectType +
+                        " transacted actions cannot be used inside a transaction block"));
```
Comment (wggihan) [inline stdlib/messaging/activemq-artemis/src/main/java/org/ballerinalang/messaging/artemis/ArtemisTransactionContext.java]: I thought we can't use outside the transaction block?

**Pull request merged**